php-checks/src/main/java/org/sonar/php/checks/FunctionDefineOutsideClassCheck.java

```diff
@@ -20,73 +20,81 @@
+public class FunctionDefineOutsideClassCheck extends PHPVisitorCheck {
 
-  private final Deque<AstNode> scopes = new ArrayDeque<>();
+  public static final String KEY = "S2007";
+  private static final String MESSAGE = "Move this %s into a class.";
+
+  private final Deque<Tree> scopes = new ArrayDeque<>();
   private final Set<String> globalVariableNames = Sets.newHashSet();
 
   @Override
-  public void init() {
-    subscribeTo(FunctionUtils.functions());
-    subscribeTo(PHPGrammar.ASSIGNMENT_EXPR);
+  public void visitScript(ScriptTree tree) {
+    scopes.clear();
+    globalVariableNames.clear();
+    super.visitScript(tree);
   }
 
   @Override
-  public void leaveFile(AstNode astNode) {
-    scopes.clear();
-    globalVariableNames.clear();
+  public void visitFunctionExpression(FunctionExpressionTree tree) {
+    scopes.push(tree);
+    super.visitFunctionExpression(tree);
+    scopes.pop();
   }
 
   @Override
-  public void visitNode(AstNode astNode) {
-    if (astNode.is(PHPGrammar.FUNCTION_DECLARATION)) {
-      getContext().createLineViolation(this, "Move this function into a class.", astNode);
-    } else if (scopes.isEmpty()) {
-      checkAssignment(astNode);
-    }
-    if (astNode.is(FunctionUtils.functions())) {
-      scopes.push(astNode);
-    }
+  public void visitMethodDeclaration(MethodDeclarationTree tree) {
+    scopes.push(tree);
+    super.visitMethodDeclaration(tree);
+    scopes.pop();
   }
 
   @Override
-  public void leaveNode(AstNode astNode) {
-    if (astNode.is(FunctionUtils.functions())) {
-      scopes.pop();
-    }
+  public void visitFunctionDeclaration(FunctionDeclarationTree tree) {
+    context().newIssue(KEY, String.format(MESSAGE, "function")).tree(tree);
+
+    scopes.push(tree);
+    super.visitFunctionDeclaration(tree);
+    scopes.pop();
   }
 
-  private void checkAssignment(AstNode astNode) {
-    AstNode memberExpr = astNode.getFirstChild();
-    AstNode variable = memberExpr.getFirstChild();
-    if (memberExpr.getNumberOfChildren() == 1 && variable.is(PHPGrammar.VARIABLE_WITHOUT_OBJECTS)) {
-      String varName = variable.getTokenOriginalValue();
-      if (memberExpr.getTokens().size() == 1 && !CheckUtils.isSuperGlobal(varName) && !globalVariableNames.contains(varName)) {
-        getContext().createLineViolation(this, "Move this variable into a class.", astNode);
-        globalVariableNames.add(varName);
+  @Override
+  public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+    super.visitAssignmentExpression(tree);
+
+    if (scopes.isEmpty() && tree.is(Kind.ASSIGNMENT) && tree.variable().is(Kind.VARIABLE_IDENTIFIER)) {
+      String varName = ((VariableIdentifierTree) tree.variable()).variableExpression().text();
+
+      if (!CheckUtils.isSuperGlobal(varName) && !globalVariableNames.contains(varName)){
+          context().newIssue(KEY, String.format(MESSAGE, "variable")).tree(tree);
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this line start at column 9. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AIndentationCheck)